python/delta_sharing/reader.py

```diff
@@ -71,19 +79,112 @@ def limit(self, limit: Optional[int]) -> "DeltaSharingReader":
         response = self._rest_client.list_files_in_table(
             self._table,
             predicateHints=self._predicateHints,
             limitHint=self._limit,
             version=self._version,
-            timestamp=self._timestamp
+            timestamp=self._timestamp,
         )
 
+        return response
+
+    @staticmethod
+    def _to_pyarrow_dataset(
+        file_urls: Union[Sequence[str], str], pyarrow_ds_options: Optional[Mapping[str, Any]] = None
+    ) -> PyArrowDataset:
+        if pyarrow_ds_options is None:
+            pyarrow_ds_options = {}
+
+        if file_urls and isinstance(file_urls, list):
+            first_url = urlparse(file_urls[0])
+        elif file_urls and isinstance(file_urls, str):
+            first_url = urlparse(file_urls)
+        else:
+            raise ValueError(
+                "PyArrow Dataset can only be instantiated from a path of a list of paths."
+            )
+
+        first_scheme = first_url.scheme
+
+        if "storage.googleapis.com" in (first_url.netloc.lower()):
+            # Apply the yarl patch for GCS pre-signed urls
+            import delta_sharing._yarl_patch  # noqa: F401
+
+        # Take URL of first file from response and assert that all
+        # the other URLs have same scheme as well.
+        # Otherwise, fsspec filesystem cannot be instantiated.
+        if file_urls and isinstance(file_urls, list):
+            same_scheme = True
+            for url in file_urls:
+                if not urlparse(url).scheme == first_scheme:
+                    same_scheme = False
+                    break
+
+            assert same_scheme, "All files did not follow the same URL scheme."
+
+        fs = fsspec.filesystem(first_scheme)
+        ds = dataset(source=file_urls, format="parquet", filesystem=fs, **pyarrow_ds_options)
+
+        return ds
+
+    def to_pyarrow_table(
+        self,
+        pyarrow_ds_options: Optional[Mapping[str, Any]] = None,
+        pyarrow_tbl_options: Optional[Mapping[str, Any]] = None,
+    ) -> PyArrowTable:
+        if pyarrow_tbl_options is None:
+            pyarrow_tbl_options = {}
+
+        response = self._get_response()
+        schema_json = loads(response.metadata.schema_string)
+
+        if len(response.add_files) == 0 or self._limit == 0:
+            return get_empty_pyarrow_table(schema_json)
+
+        file_urls = [f.url for f in response.add_files]
+
+        if self._limit is None:
+            ds = DeltaSharingReader._to_pyarrow_dataset(file_urls, pyarrow_ds_options)
+            return ds.to_table(**pyarrow_tbl_options)
+        else:
+            left = self._limit
+            pa_tables = []
+
+            for file_url in file_urls:
+                ds = DeltaSharingReader._to_pyarrow_dataset(file_url, pyarrow_ds_options)
+                tbl: PyArrowTable = ds.head(left, **pyarrow_tbl_options)
+                pa_tables.append(tbl)
+                left -= tbl.num_rows
+                assert (
```

Comment:
is this a hard limit? and does it require exact `limit` number of rows returned?

Reply:
yes it results exactly the number of rows asked for. but does it file by file instead as I saw that in practice this is faster than just calling `.head()` on the pyarrow table.

Reply:
right. I wonder we don't have to fail if we got a few more rows?